core/templates/services/debug-info-tracker.service.ts

```diff
@@ -23,6 +23,8 @@ import { Injectable } from '@angular/core';
   providedIn: 'root'
 })
 export class DebugInfoTrackerService {
+  // The type of an action is not fixed. It could be anything so it is defined
+  // as any.
```

Comment:
Can't we change this to Object as well in that case? Also, how did you come to the conclusion that type of action is not fixed? Can you please explain what an action is and give some examples as well?

Reply:
We can, but if we in future if we access this variable ` _sequenceOfActions` and use one action like
```javascript
action = _sequenceOfActions[0];
console.log(action.length);
```
It would show an type error as `action` is of type `Object` and it has no property `length`.

Reply:
Examples where this service is used 
- [Example 1](https://github.com/oppia/oppia/blob/ae649aa08f1375457ec9e3c90257197b68fec7cd/core/templates/pages/exploration-editor-page/preview-tab/preview-tab.directive.ts#L144)
- [Example 2](https://github.com/oppia/oppia/blob/ae649aa08f1375457ec9e3c90257197b68fec7cd/core/templates/pages/exploration-player-page/learner-experience/conversation-skin.directive.ts#L1035)

Reply:
There are only two types of action being added based on the usage:

1. Simple string
2. Object of format `{interactionId: string, cardIndex: number}`

So, we have to add an interface which allows either of these two types.

Further the service is for tracking a learner action (as mentioned in the docstring). So, before going with this option, we should confirm what are the types of learner actions which can be performed by a user. Once we have that list, we can determine all the types of actions then.

@aks681 @seanlip any suggestions on how can we determine all the learner actions which can be performed?

Also, I think this would require more analysis and you can proceed without this change in this PR and opening a fresh one for this change.

Reply:
It just needs analysis / tracing through the code. Figure out what calls this service or adds stuff to the variable, and then enumerate the calls and see if there is anything in common.

Reply:
Also if you are planning to punt some work to a follow-up PR, please make sure there's an issue created for it, and link to it here before considering this comment addressed. Thanks!

Reply:
@seanlip I think, we should also think about the future scope of this function i.e. what else it can be used for in the future. It may be used to track only two or three types of action now, but it may be used for a different type of object in the future?

Reply:
If it is then the developer should update the types.

Your concern is valid though. I think we should perhaps have an Action frontend domain object and update the code to only use that. Everything should be clearly typed, in general, and we should not be passing around a bunch of dicts.

Reply:
@seanlip So, I think we should create a new issue about creating a separate ActionObjectFactory and to use that everywhere. And, let the type of action be any till the issue is resolved.

Reply:
Sure -- feel free to go ahead and create one, and please link to it here. Also loop in the person who originally wrote the debug-info-tracker-service to get any clarification on the requirements and to check they're happy with this.

Also, are you planning to resolve that issue? I just wanted to be clear that "fixing TS errors" also includes doing things like this (i.e. basically, do what it takes to get the outcome that's desired).

Reply:
@seanlip I just created the issue(#8877). Yup, I will resolve the issue when I'll be removing `any` from this file. But for now, I think we should not change this file then?

Reply:
Sure, sounds good, thanks!

Reply:
Created a pr to fix this.